### PR TITLE
fix: Support case insensitive reference resolution

### LINKED ISSUE #2717: KeyNotFoundException when starting stryker
**Describe the bug**
At initialization, stryker populates a dictionary of projects, all attached to a .csproj file.
When the filesystem is case insensitive, the file name on disk and the one in the sln/csproj can be different, leading the exception.

**Logs**
``` log
❯ dotnet stryker

   _____ _              _               _   _ ______ _______  
  / ____| |            | |             | \ | |  ____|__   __| 
 | (___ | |_ _ __ _   _| | _____ _ __  |  \| | |__     | |    
  \___ \| __| '__| | | | |/ / _ \ '__| | . ` |  __|    | |    
  ____) | |_| |  | |_| |   <  __/ |    | |\  | |____   | |    
 |_____/ \__|_|   \__, |_|\_\___|_| (_)|_| \_|______|  |_|    
                   __/ |                                      
                  |___/                                       


Version: 3.10.0

[18:26:58 INF] Analysis starting.
[18:26:59 INF] Identifying projects to mutate in C:\Git\<redacted>.sln. This can take a while.
[18:27:55 INF] Found 29 source projects
[18:27:55 INF] Found 6 test projects
[18:27:55 INF] Analysis complete.
[18:27:55 INF] Time Elapsed 00:00:56.4337834
Unhandled exception. System.Collections.Generic.KeyNotFoundException: The given key 'C:\Git\<redacted>.Interfaces.csproj' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Stryker.Core.Initialisation.InputFileResolver.FindDependentProjects(IReadOnlyCollection`1 projectsUnderTest) in /_/src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs:line 101
   at Stryker.Core.Initialisation.InputFileResolver.ResolveSourceProjectInfos(StrykerOptions options) in /_/src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs:line 81
   at Stryker.Core.Initialisation.InitialisationProcess.GetMutableProjectsInfo(StrykerOptions options) in /_/src/Stryker.Core/Stryker.Core/Initialisation/InitialisationProcess.cs:line 58
   at Stryker.Core.Initialisation.ProjectOrchestrator.MutateProjects(StrykerOptions options, IReporter reporters, ITestRunner runner)+MoveNext() in /_/src/Stryker.Core/Stryker.Core/Initialisation/ProjectOrchestrator.cs:line 50
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Stryker.Core.StrykerRunner.RunMutationTest(IStrykerInputs inputs, ILoggerFactory loggerFactory, IProjectOrchestrator projectOrchestrator) in /_/src/Stryker.Core/Stryker.Core/StrykerRunner.cs:line 61
   at Stryker.CLI.StrykerCli.RunStryker(IStrykerInputs inputs) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 93
   at Stryker.CLI.StrykerCli.<>c__DisplayClass10_0.<Run>b__0() in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 68
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.<>c__DisplayClass143_0.<OnExecute>b__0(CancellationToken _)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.ExecuteAsync(String[] args, CancellationToken cancellationToken)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at Stryker.CLI.StrykerCli.Run(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 74
   at Stryker.CLI.Program.Main(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/Program.cs:line 14
```

(Sorry, I can't provide a full log because of the filenames)

**Expected behavior**
Stryker should not crash :)

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Type of project: core
 - Framework Version: 7
 - Stryker Version: 3.10.0

**Additional context**
I'm willing to help and make a PR if you want me to :)

### PR DESCRIPTION
**Pull request by dupdob**
Fix #2717 + add several tests for project orchestrator class (brought in #2624 tests so they are not stuck due to hard decisions)

### CODE REVIEW

**Comment by rouke-broersma** [comment]
awesome :)

File: src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/ProjectOrchestratorTests.cs [hunk 1]
```diff
@@ -8,154 +8,439 @@
...
+                    , new[] {libraryProject}).Object },
+                { "MyProject.UnitTests", TestProjectAnalyzerMock(testCsprojPathName, csprojPathName).Object }
+            };
+            var target = BuildProjectOrchestrator(analyzerResults, out var mockRunner);
 
+            _fileSystem.Directory.SetCurrentDirectory(_fileSystem.Path.GetFullPath(testCsprojPathName));
+            // act
+            var result = target.MutateProjects(options, _reporterMock.Object, mockRunner.Object).ToList();
+            // assert
+            result.Count.ShouldBe(1);
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/ProjectOrchestratorTests.cs:265]:
```suggestion
            result.ShouldBeHaveSingleItem();
```

**Review by richardwerkman — COMMENTED** [review]
I see @rouke-broersma already merged the PR, but I was still in progress of reviewing. I have some suggestions, maybe they can be adressed in a seperate PR? Thanks for the extra test coverage :) Awesome work

File: src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs [hunk 2]
```diff
@@ -98,7 +98,19 @@ public IReadOnlyCollection<SourceProjectInfo> ResolveSourceProjectInfos(StrykerO
...
+                    if (!dependents.ContainsKey(reference))
+                    {
+                        // find a match disregarding case
+                        correctReference = dependents.Keys.FirstOrDefault(r => string.Compare(r, reference, true, System.Globalization.CultureInfo.InvariantCulture) == 0);
+                        if (correctReference == null)
+                        {
+                            _logger.LogWarning("Project {0} depends upon unknown projects {1}. Compilation may fail.", result.ProjectFilePath, reference);
+                            continue;
+                        }
+                        _logger.LogWarning("Project {0} depends upon unknown projects {1} which should have been cased like this {2}.", result.ProjectFilePath, reference, correctReference);
```
[the remark this answers is not in the document]
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs:111]: I think this should be seen more as a tolerance than as a supported behavior. There is no valid use case (that I am aware of) that explain why there is a case difference between the reference declared and the actual project filename.
VsStudio tolerates this, so we might as well. 
That being said, this is undocumented territory and there is no guarantee everything will be fine afterward: there may be consequences.

Plus, fixing the cause is simple: either recreate the dependence or align name and filename.

**Pull request merged**